---
title: "matrix-eQTL"
author: "em"
date: "December 28, 2019"
output:
  html_document:
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
setwd('~/Documents/capsella-eqtl/')
library(MatrixEQTL)
```

##Get the batch-corrected expression data together for mapping
```{r, eval=F}
load('data/wgcna-files/combatOutput.rda')

mycbnames = read.table("data/wgcna-files/mycb-names", stringsAsFactors = F)

qn <- function(x){qqnorm(x, plot.it=FALSE)$x}
# nnormed = data.frame(t(apply(outtable[,-1], 1,qn)))
normcb = apply(mycb, 1, qn)#quantile normalize, now columns genes, rows are inds


mycbdf = data.frame(id = mycbnames$x, normcb, stringsAsFactors = F)

matrixeqtlids = data.frame(V1 = strsplit('10    100    101    102_long    105_clipinplace_sorted    106s_clipped    107    108_3    109    110    111    112    113    114_long    115    117s_clipped    118    11s_clipped    121    123    124_3    125_clipped.bam    126s_clipped    128    129s_clipped    13    131s_clipped    132    133    135    136    137s_clipped    138    139    140    141    142    143s_clipped    144s_clipped    147    148    14_clipped.bam    15    151    152    153    154    155_clipped.bam    156    157    158    160    161_clipped.bam    162    163_clipped    165_clipped.bam    167    16_clipped    17    170_clipped    174_clipped.bam    175_clipped    176    177    178    18    181    182_3    183    184_3    186    187    189    190    192    193    194    195    197    198    1_clipped    20    200    202_long    203_long    204_long    23s_clipped    24    25    26    27    28    29_clipped    30    31    33    34    35    36    38    39s_clipped    41    43s_clipped    44s_clipped    47    49_clipped.bam    4s_clipped    50_long    52    54_clipped    55_clipped    58    5_clipped    61    63    64_replacement    65_clipped    66_clipped    67    70s_clipped    71    72    74    75    76    78s_clipped    79    7s_clipped    8    80s_clipped    81s_clipped    82s_clipped    83    85_long    86    89    9    91    92    93    94    95    96    97_3    98_clipped    99_clipped', split = "    ")[[1]], stringsAsFactors = F)



meExp = data.frame(c('id', row.names(mycb)),t(dplyr::left_join(matrixeqtlids, mycbdf, by = c('V1'='id'))), stringsAsFactors = F)
write.table(meExp, file = 'data/matrixeqtl-files/cbExp', row.names=F, quote=F, col.names=F)

```

```{bash, eval=F, echo=F}
rsync data/matrixeqtl-files/cbExp josep993@rsync.hpcc.msu.edu:trans-eQTL/data/

rsync -ave ssh josep993@rsync.hpcc.msu.edu:trans-eQTL/bin/matrixeqtl-batched.rda data/matrixeqtl-files
```


```{r}
load('data/matrixeqtl-files/matrixeqtl-batched.rda')
plot(meCovar)

geneloc = read.table('data/matrixeqtl-files/genelocation.matrixeqtl', header=T, stringsAsFactors = F)
geneloc$gene = as.character(geneloc$geneid)

## merge gene locations with eQTL results (looking at Tag sites only)
locmerge = dplyr::left_join(meCovar$all$eqtls, geneloc, by = "gene")
locmerge$snpchr = sapply(locmerge$snps, function(x){strsplit(as.character(x),':')[[1]][1]})
locmerge$snppos = sapply(locmerge$snps, function(x){strsplit(as.character(x),':')[[1]][2]})


#get xlim for the plot
myxlim = lapply(1:8, function(x){
  mychr = dplyr::filter(locmerge, snpchr == paste('scaffold_',as.character(x),sep=""))
  mymax = max(as.numeric(mychr$snppos))
  mymin = min(as.numeric(mychr$snppos))
  return(c(mymin, mymax))
  })


##remove ones with >100 eQTLs
genecounts = dplyr::count(locmerge, gene)
names(genecounts) = c('gene','numbereQTLs')

par(mfrow=c(1,1), mar=c(5,5,2,2))
hist(genecounts$numbereQTLs, col="navy", border="navy", breaks=100, main = "eQTLs per gene", xlab ="")
hist(genecounts$numbereQTLs, col="navy", border="navy", breaks=100, main = "eQTLs per gene", xlab ="", xlim = c(50,1000), ylim =c(0,80))
weirdos = dplyr::filter(genecounts, numbereQTLs>29)
dim(weirdos)
keepsnps = dplyr::filter(genecounts, numbereQTLs < 30)

finalset = dplyr::filter(locmerge, gene %in% keepsnps$gene, !is.na(chr))

eqtlsforplot = dplyr::filter(finalset, FDR < 0.1)
nrow(eqtlsforplot)

#postscript("figures/eqtls-pcs.eps",height=9,width=9,paper="special",horizontal=FALSE,colormodel="cymk")
par(mfrow=c(8,8), mar=c(0,0,0,0))
par(mar=c(0,0,0,0))
layout(matrix(c(1,2:9,1,10:17,1,18:25,1,26:33,1,34:41,1,42:49,1,50:57,1,58:65, rep(66,9)), 9, 9, byrow = TRUE))
plot.new()
text(0.25,0.5,"Gene location", srt=90, cex=2.5)
for (i in 8:1){ ##gene locations
  for (j in 1:8){  ##snp location
  mygenechr = paste('scaffold_',as.character(i),sep="")
  mysnpchr = paste('scaffold_',as.character(j),sep="")
mychrom = dplyr::filter(eqtlsforplot, chr ==mygenechr & snpchr == mysnpchr)
mysigfdr = dplyr::filter(mychrom, FDR<0.1)
plot(mychrom$snppos,mychrom$s1, col = 'darkgray', xlab = "", ylab = "", yaxt="n", xaxt="n", lwd=2,cex=1, xlim = myxlim[[j]])
#text(100000,100000,paste('gene location is',as.character(i)))
if (j==1){mtext(i, side = 2, cex=1.5, las=2, line=1)} ##if at chrom 1 for snps
if(i==1){mtext(j, side=1, cex=1.5, line=1)} ## if at chrom 1 for genes

  }
}
plot.new()
text(0.5,0.25, "SNP location", cex=2.5)




```

Cis vs trans effect sizes?
```{r}
### get the distances of eqtls from the gene

finalset$qtldist = apply(finalset, 1, function(x){
  qtldist = 1e8
  if (x[8] == x[11]){ ##on the same chromosome
  qtldist1 = abs(as.numeric(x[9])-as.numeric(x[12])) ##gene start - snpchr 
  qtldist2 = abs(as.numeric(x[10])-as.numeric(x[12])) ## gene end - snp chr
  qtldist = min(qtldist1, qtldist2)
  }
  return(qtldist)
})

hist(finalset$qtldist, xlim = c(0,1e4), col="navy", main = "distances on same chrom", breaks=100000, ylim = c(0,6000))

ciseqtls = dplyr::filter(finalset, qtldist < 5000, FDR < 0.1)
transeqtls = dplyr::filter(finalset, qtldist > 5000, FDR < 0.1)
nrow(ciseqtls)
nrow(transeqtls)
## plot of how many cis and trans eqtls there are at FDR = 0.1
mybp= barplot(c(nrow(ciseqtls), nrow(transeqtls)), xaxt='n', col = "darkgray", border="white", ylim = c(0,5000), xlim = c(-0.5,3), yaxt="n")
axis(1, at = mybp, labels = c('cis','trans'), cex.axis=2, tick = F, col = "white")
axis(2, las=2, cex.axis=2)

### Compare effect sizes of cis and trans eqtls.
mytest = t.test(abs(ciseqtls$beta), abs(transeqtls$beta)) ## cis eqtls have smaller effects but not by that much
mytest
mytest$p.value

par(mar=c(7,7,2,2), mfrow=c(1,1))
plot(jitter(c(rep(1, nrow(ciseqtls)), rep(2, nrow(transeqtls)))), abs(c(ciseqtls$beta, transeqtls$beta)), bty="n", xlim = c(0.5,2.5), xaxt="n", yaxt = "n", col = "gray", xlab = "", ylab = "")
axis(2, las=2, cex.axis=2)
mtext('abs(beta)',2, line=5, cex=2)
axis(1,at = c(1,2), lab = c('cis','trans'), cex.axis=2)
points(c(1,2), c(mean(abs(ciseqtls$beta)), mean(abs(transeqtls$beta))), cex=3, pch=16)
```

Compare allele freqs of cis and trans
```{r}
finalset$rs = as.character(finalset$snps)
mygemma = read.table('data/gemma-files/tb.16.assoc.txt', header=T, stringsAsFactors=F)##read in mafs from a gemma file
finalaf = dplyr::left_join(finalset, mygemma[,c(2,5,6,7)], by = 'rs')

ciseqtls = dplyr::filter(finalaf, qtldist < 5000, FDR < 0.1)
transeqtls = dplyr::filter(finalaf, qtldist > 5000, FDR < 0.1)


t.test(ciseqtls$af, transeqtls$af) ## cis eqtls have smaller effects but not by that much
t.test(ciseqtls$af, transeqtls$af)$p.value


par(mar=c(7,7,2,2), mfrow=c(1,1))
plot(jitter(c(rep(1, nrow(ciseqtls)), rep(2, nrow(transeqtls)))), c(ciseqtls$af, transeqtls$af), bty="n", xlim = c(0.5,2.5), xaxt="n", yaxt = "n", col = "gray", xlab = "", ylab = "")
axis(2, las=2, cex.axis=2)
mtext('MAF',2, line=5, cex=2)
axis(1,at = c(1,2), lab = c('cis','trans'), cex.axis=2)
points(c(1,2), c(mean(ciseqtls$af, na.rm=T), mean(transeqtls$af, na.rm=T)), cex=3, pch=16)
##driven by the hotspot? remove duplicates from the trans
transunique = transeqtls[!duplicated(transeqtls$rs),]
dim(transunique)
cisunique = ciseqtls[!duplicated(ciseqtls$rs),]
dim(cisunique)

t.test(cisunique$af, transunique$af) ## cis eqtls have smaller effects but not by that much

par(mar=c(7,7,2,2), mfrow=c(1,1))
plot(jitter(c(rep(1, nrow(cisunique)), rep(2, nrow(transunique)))), c(cisunique$af, transunique$af), bty="n", xlim = c(0.5,2.5), xaxt="n", yaxt = "n", col = "gray", xlab = "", ylab = "")
axis(2, las=2, cex.axis=2)
mtext('MAF',2, line=5, cex=2)
axis(1,at = c(1,2), lab = c('cis','trans'), cex.axis=2)
points(c(1,2), c(mean(cisunique$af, na.rm=T), mean(transunique$af, na.rm=T)), cex=3, pch=16)

save('finalaf', file = "data/allbyall.rda")

```

Remake the plot of all eQTLs but add in the coexpression eqtls

```{r}
table1 = read.table('Table1', header=T, stringsAsFactors=F)
table1$scaf = c(7,8,4,4)
table1$pos = sapply(table1$loc, function(x){strsplit(x, split=":")[[1]][2]})

#postscript("figures/eqtls-pcs.eps",height=9,width=9,paper="special",horizontal=FALSE,colormodel="cymk")
par(mfrow=c(8,8), mar=c(0,0,0,0))
layout(matrix(c(1,2:9,1,10:17,1,18:25,1,26:33,1,34:41,1,42:49,1,50:57,1,58:65, rep(66,9)), 9, 9, byrow = TRUE))
plot.new()
text(0.25,0.5,"Gene location", srt=90, cex=2.5)
for (i in 8:1){ ##gene locations
  for (j in 1:8){  ##snp location
  mygenechr = paste('scaffold_',as.character(i),sep="")
  mysnpchr = paste('scaffold_',as.character(j),sep="")
mychrom = dplyr::filter(eqtlsforplot, chr ==mygenechr & snpchr == mysnpchr)
plot(mychrom$snppos,mychrom$s1, col = 'darkgray', xlab = "", ylab = "", yaxt="n", xaxt="n", lwd=2,cex=1, xlim = myxlim[[j]])
mycoexp = dplyr::filter(table1, scaf == j)
if (nrow(mycoexp)>0){
  mycoexpoverlaps = dplyr::filter(mychrom, snppos %in% mycoexp$pos)
  points(mycoexpoverlaps$snppos, mycoexpoverlaps$s1, col = "black", pch=16)
}


#text(100000,100000,paste('gene location is',as.character(i)))
if (j==1){mtext(i, side = 2, cex=1.5, las=2, line=1)} ##if at chrom 1 for snps
if(i==1){mtext(j, side=1, cex=1.5, line=1)} ## if at chrom 1 for genes


  }
}
plot.new()
text(0.5,0.25, "SNP location", cex=2.5)
```


